src/metrics.rs

```diff
@@ -50,19 +72,56 @@ pub fn init() -> crate::Result<()> {
         registry: Arc::clone(&registry),
         cardinality_counter: Arc::clone(&cardinality_counter),
     };
-    // Apply a layer to capture tracing span fields as labels.
-    let recorder = TracingContextLayer::new(VectorLabelFilter).layer(recorder);
+
+    // If enabled, apply a layer to capture tracing span fields as labels.
+    let recorder: Box<dyn metrics::Recorder> = if tracing_context_layer_enabled() {
+        Box::new(TracingContextLayer::new(VectorLabelFilter).layer(recorder))
+    } else {
+        Box::new(recorder)
+    };
+
     // Register the recorder globally.
-    metrics::set_boxed_recorder(Box::new(recorder)).map_err(|_| "recorder already initialized")?;
+    metrics::set_boxed_recorder(recorder).map_err(|_| "recorder already initialized")?;
 
     // Done.
     Ok(())
 }
 
+/// [`VectorRegistry`] is a vendored version of [`metrics_util::Registry`].
```

Comment:
What are the changes that required us to vendor? Could we add that as comments here? It'd be nice to contribute them back upstream if it makes sense, too.

We may need to copy their license / attribution over here too.

Reply:
We are removing the generational wrappers that upstream added, as they might've been the cause of the performance issues under concurrency. They were added since I originally implemented the `Registry` code, I didn't want to bother with updating my mental model with this new piece at the time.

I have a suspicion that the atomics usage in the generational has a change to cause constant cache invalidation at some scenario - however, it's based on my empiric observation, rather than a comprehensive mental model of the behaviour to back it up - so I decided we could just eliminate them - for now. Maybe in the long term too - we don't need them, so why pay the price. They're not zero-cost.

Reply:
Added at 0008ccc54b443edb1bd5df492a48bb2f439aff6b